File: bms/go/geo/geo_s2_CellID_String/geo_s2_CellID_String.go
package test

import (
	"bytes"
	"math/bits"
	"strconv"
)

const (
	geo_s2_NumFaces = 6
	geo_s2_PosBits  = 2*geo_s2_MaxLevel + 1
	geo_s2_MaxLevel = 30
)

func (ci geo_s2_CellID) String() string {
	if !ci.IsValid() {
		return "Invalid: " + strconv.FormatInt(int64(ci), 16)
	}
	var b bytes.Buffer
	b.WriteByte("012345"[ci.Face()]) // values > 5 will have been picked off by !IsValid above
	b.WriteByte('/')
	for level := 1; level <= ci.Level(); level++ {
		b.WriteByte("0123"[ci.ChildPosition(level)])
	}
	return b.String()
}

func (ci geo_s2_CellID) IsValid() bool {
	return ci.Face() < geo_s2_NumFaces && (ci.lsb()&0x1555555555555555 != 0)
}

func (ci geo_s2_CellID) Face() int { return int(uint64(ci) >> geo_s2_PosBits) }

func (ci geo_s2_CellID) lsb() uint64 { return uint64(ci) & -uint64(ci) }

func (ci geo_s2_CellID) ChildPosition(level int) int {
	return int(uint64(ci)>>uint64(2*(geo_s2_MaxLevel-level)+1)) & 3
}

func (ci geo_s2_CellID) Level() int {
	return geo_s2_MaxLevel - geo_s2_findLSBSetNonZero64(uint64(ci))>>1
}

func geo_s2_findLSBSetNonZero64(x uint64) int {
	if x == 0 {
		return 0
	}
	return bits.TrailingZeros64(x)
}

type geo_s2_CellID uint64

File: translations_only/fail/go/geo/geo_s2_CellID_String/gemini/CAPR/geo_s2_CellID_String.rs


const GEO_S2_NUM_FACES: u8 = 6;
const GEO_S2_POS_BITS: u8 = 2 * GEO_S2_MAX_LEVEL + 1;
const GEO_S2_MAX_LEVEL: u8 = 30;

#[derive(Copy, Clone, Debug)]
pub struct GeoS2CellID(pub u64);

impl GeoS2CellID {
    pub fn is_valid(&self) -> bool {
        self.0 >> GEO_S2_POS_BITS < GEO_S2_NUM_FACES as u64
            && (self.0 & 0x1555555555555555 != 0)
    }

    pub fn face(&self) -> u8 {
        (self.0 >> GEO_S2_POS_BITS) as u8
    }

    pub fn lsb(&self) -> u64 {
        self.0 & !self.0
    }

    pub fn child_position(&self, level: u8) -> u8 {
        ((self.0 >> (2 * (GEO_S2_MAX_LEVEL - level) + 1)) & 3) as u8
    }

    pub fn level(&self) -> u8 {
        GEO_S2_MAX_LEVEL - (geo_s2_find_lsb_set_non_zero_64(self.0) >> 1)
    }

    pub fn to_string(&self) -> String {
        if !self.is_valid() {
            return format!("Invalid: {:x}", self.0);
        }
        let mut buf = String::new();
        buf.push_str(&"012345"[self.face() as usize..self.face() as usize + 1]);
        buf.push('/');
        for level in 1..=self.level() {
            buf.push_str(&"0123"[self.child_position(level) as usize..self.child_position(level) as usize + 1]);
        }
        buf
    }
}

fn geo_s2_find_lsb_set_non_zero_64(x: u64) -> u8 {
    if x == 0 {
        return 0;
    }
    x.trailing_zeros() as u8
}

